### ISSUE TITLE
Easier way to filter parameters by Module type

### ISSUE THREAD

**Issue by rkaplan** [pytorch/pytorch#15067]
It is relatively common now to avoid applying weight decay to BatchNorm parameters. The recommended way to do this in PyTorch (https://github.com/pytorch/pytorch/issues/1402) is to use named_parameters and filter by the module name string. However, this is quite brittle. For example, I use a helper function like:

```
def CBR(nin, nout, filter_size, stride=1, inplace=True):
    """ Stack of Conv-Batchnorm-ReLU """
    pad = int((filter_size - 1)/2)
    return nn.Sequential(
        nn.Conv2d(nin, nout, filter_size, stride, pad, bias=False),
        nn.BatchNorm2d(nout),
        nn.ReLU(inplace=inplace))
```

thoughout my code to build stacks of conv / bn / ReLU blocks. The nn.Sequential hides the names of each module, so all I can see when iterating through `.named_parameters()` is `['l1.0.weight', 'l1.1.weight', 'l1.1.bias']`. There's no way to tell which parameters are BatchNorm params.

My options seem to be:
1) Refactor code to never use BatchNorm layers inside sequentials (not practical for maintainability)
2) When passing parameters to the optimizer, iterate over modules first. Something like:
```
params = [m.parameters() for m in model.modules() if isinstance(m, torch.nn.BatchNorm2d)]
optimizer = torch.optim.SGD(itertools.chain(*params))
```
That would be fine to do in my function that builds the optimizer, except it makes it a pain for callers that want to do fine-tuning on only a subset of parameters.

It feels like the root issue is that parameters don't have a reference to the modules that they belong to. Why is that? Would it be feasible to add another field to Parameters to include this reference? Then it would be as simple as: `[p for p in parameters if isinstance(p.module, nn.BatchNorm2d)]`.

**Comment by rkaplan** [comment]
Realized two things:

1. Users can resolve this with the existing API by building a parameter -> module mapping manually using `recurse=False`, i.e.:
```
        param_to_module = {}
        for m in model.modules():
            for p in m.parameters(recurse=False):
                param_to_module[p] = m
```

2. In general, parameters might be able to belong to more than one module if there is parameter sharing going on, so perhaps a `.module` member of Parameters isn't a great idea. Hence going to closing this, but still curious if a more direct way to access the corresponding module(s?) of a parameter will be supported in the future.

**Comment by askerlee** [comment]
How about mapping each param to its module name? I guess that's enough for most scenarios.
